File: python3/combinations.py
from itertools import combinations

# Space     : O(n^2)
# Time      : O(n!)

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        arr = [i for i in range(1, n+1)]
        res = []
        
        comb = combinations(arr, k)
        for c in comb:
            res.append(list(c))
        
        return res

File: golang/combinations.go
package main

/*
Space	: O(k)
Time	: O(k^n)
*/

func combine(n int, k int) [][]int {
	ans := [][]int{}
	path := []int{}
	var getCombination func(int, int)

	getCombination = func(idx, k int) {
		if k == 0 {
			ans = append(ans, append([]int{}, path...))
			return
		}

		for i := idx; i <= n; i++ {
			path = append(path, i)
			getCombination(i+1, k-1)
			path = path[:len(path)-1]
		}
	}

	getCombination(1, k)
	return ans
}
